File: trip-kit/src/main/java/com/skedgo/android/tripkit/UberBookingResolver.java
package com.skedgo.android.tripkit;

import android.content.Intent;
import android.content.res.Resources;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import rx.Observable;
import rx.functions.Func1;

final class UberBookingResolver implements BookingResolver {
  private static final String UBER_PACKAGE = "com.ubercab";
  private final Resources resources;
  private final Func1<String, Boolean> isPackageInstalled;

  public UberBookingResolver(
      @NonNull Resources resources,
      @NonNull Func1<String, Boolean> isPackageInstalled) {
    this.resources = resources;
    this.isPackageInstalled = isPackageInstalled;
  }

  @Override public Observable<BookingAction> performExternalActionAsync(ExternalActionParams params) {
    final BookingAction.Builder actionBuilder = BookingAction.builder();
    actionBuilder.bookingProvider(BookingResolver.UBER);
    if (isPackageInstalled.call(UBER_PACKAGE)) {
      final BookingAction action = actionBuilder.hasApp(true).data(
          new Intent(Intent.ACTION_VIEW).setData(Uri.parse("uber://"))
      ).build();
      return Observable.just(action);
    } else {
      final BookingAction action = actionBuilder.hasApp(false).data(
          new Intent(Intent.ACTION_VIEW).setData(Uri.parse("https://m.uber.com/sign-up"))
      ).build();
      return Observable.just(action);
    }
  }

  @Nullable @Override public String getTitleForExternalAction(String externalAction) {
    return isPackageInstalled.call(UBER_PACKAGE)
        ? resources.getString(R.string.open_uber)
        : resources.getString(R.string.get_uber);
  }
}

File: TripKitAndroid/src/main/java/com/skedgo/android/tripkit/bookingproviders/UberBookingResolver.kt
package com.skedgo.android.tripkit.bookingproviders

import android.content.Intent
import android.net.Uri
import com.skedgo.android.tripkit.BookingAction
import com.skedgo.android.tripkit.ExternalActionParams
import rx.Observable
import rx.functions.Func1

internal const val UBER_PACKAGE = "com.ubercab"

class UberBookingResolver(
    private val isPackageInstalled: Func1<String, Boolean>,
    private val getAppIntent: Func1<String, Intent>
) : BookingResolver {
  override fun performExternalActionAsync(params: ExternalActionParams): Observable<BookingAction> {
    val actionBuilder = BookingAction.builder()
    actionBuilder.bookingProvider(BookingResolver.UBER)

    if (isPackageInstalled.call(UBER_PACKAGE)) {
      val intent = getAppIntent.call(UBER_PACKAGE)
      val segment = params.segment()

      intent.data = Uri.parse("uber://?action=setPickup" +
          "&pickup[latitude]=${segment.from.lat}" +
          "&pickup[longitude]=${segment.from.lon}" +
          "&dropoff[latitude]=${segment.to.lat}" +
          "&dropoff[longitude]=${segment.to.lon}")

      val action = actionBuilder.hasApp(true).data(intent).build()
      return Observable.just(action)
    } else {
      val data = Intent(Intent.ACTION_VIEW)
          .setData(Uri.parse("https://play.google.com/store/apps/details?id=" + UBER_PACKAGE))
      val action = actionBuilder
          .hasApp(false)
          .data(data)
          .build()
      return Observable.just(action)
    }
  }

  override fun getTitleForExternalAction(externalAction: String): String? = null
}